RFC0022.md

```diff
@@ -0,0 +1,135 @@
+ * code to the caller.
+ *
+ * By default, the spawned processes will be PMIx "connected" to
+ * the parent process upon successful launch (see PMIx_Connect
+ * description for details). Note that this only means that the
+ * parent process (a) will be given a copy of the  new job's
+ * information so it can query job-level info without
+ * incurring any communication penalties, and (b) will receive
+ * notification of errors from process in the child job.
+ *
+ * Job-level directives can be specified in the job_info array. This
+ * can include:
+ *
+ * (a) PMIX_NON_PMI - processes in the spawned job will
+ *     not be calling PMIx_Init
+ *
+ * (b) PMIX_TIMEOUT - declare the spawn as having failed if the launched
+ *     procs do not call PMIx_Init within the specified time
+ *
+ * (c) PMIX_NOTIFY_COMPLETION - notify the parent process when the
+ *     child job terminates, either normally or with error
+ */
+PMIX_EXPORT pmix_status_t PMIx_Spawn(const pmix_info_t job_info[], size_t ninfo,
+                                     const pmix_app_t apps[], size_t napps,
+                                     char nspace[]);
+```
+In this definition, the job\_info array includes directives passed to the launcher to direct its behavior. The pmix\_app\_t array contains one or more of the following structures:
+```c
+/****    PMIX APP STRUCT    ****/
+typedef struct pmix_app {
+    char *cmd;
+    char **argv;
+    char **env;
+    char *cwd;
+    int maxprocs;
+    pmix_info_t *info;
+    size_t ninfo;
+} pmix_app_t;
+```
+The argv array contains the command line arguments of the application itself, while the env array contains environmental variables to be set prior to execution of application processes.
+
+#### Modifying Environmental Variables
+It is sometimes desirable or required that standard environmental variables (e.g., PATH, LD\_LIBRARY\_PATH, LD\_PRELOAD) be modified prior to executing an application binary or a starter such as mpiexec - this is particularly true when tools/debuggers are used to start the application. This RFC proposes the definition of a new PMIx structure and associated attributes for specifying such operations:
+```c
+/* Provide a structure for specifying environment variable modifications
+ * Standard environment variables (e.g., PATH, LD_LIBRARY_PATH, and LD_PRELOAD)
+ * take multiple arguments separated by delimiters. Unfortunately, the delimiters
+ * depend upon the variable itself - some use semi-colons, some colons, etc. Thus,
+ * the operation requires not only the name of the variable to be modified and
+ * the value to be inserted, but also the separator to be used when composing
+ * the aggregate value
+ */
+typedef struct {
+    char *envar;
+    char *value;
+    char separator;
+} pmix_envar_t;
+
+/* environmental variable operation attributes */
+#define PMIX_SET_ENVAR          "pmix.envar.set"          // (pmix_envar_t*) set the envar to the given value,
```

Comment:
If we model this after `setenv` then we need a `PMIX_SET_ENVAR` (which overwrites if pre-exiting value exists) and `PMIX_SET_ENVAR_NO_OVERWRITE` (which does not).

Reply:
Missed `PMIX_ADD_ENVAR` is the 'no overwrite' version. So nevermind that comment.